### PR TITLE
Add space engineers

### LINKED ISSUE #483: Space Engineers Dedicated Server Egg
This server needs mono or wine to work properly. It allredy exists a egg with this logic.
Link to the egg with mono logic: https://github.com/parkervcp/eggs/tree/master/steamcmd_servers/eco

Service: Space Engineers Dedicated Server
Steam App ID: 298740

Does this expand an already existing service: N

Link to game: https://www.spaceengineersgame.com

Links for server downloads: https://www.spaceengineersgame.com/dedicated-servers.html

Links for install steps/docs: https://www.spaceengineersgame.com/dedicated-servers.html

### PR DESCRIPTION
**Pull request by TuEye**
Many parts of this egg are based on the work of mmmaxwwwell (https://github.com/mmmaxwwwell/space-engineers-dedicated-docker-linux). So many thanks to him.

Should resolve #483

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Did you branch your changes and PR from that branch and not from your master branch?

### New Server Submissions:

1. [x] Does your submission pass tests (server is connectable)?
2. [x] Have you added the server to the main README.md?
3. [x] Have you added a unique README.md for the server you are adding?

### CODE REVIEW

File: steamcmd_servers/space_engineers/egg-space-engineers.json [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    "exported_at": "2021-03-11T23:55:52+01:00",
+    "name": "Space Engineers",
+    "author": "tueye@tuworld.de",
+    "description": "Space Engineers is a voxel-based sandbox game set in space and on planets.",
+    "features": null,
+    "images": [
+        "quay.io\/parkervcp\/pterodactyl-images:debian_wine-5"
+    ],
+    "file_denylist": [],
+    "startup": ".\/start.sh",
```
Comment (parkervcp) [inline steamcmd_servers/space_engineers/egg-space-engineers.json:16]: This is a no from me for having it as a start script. Users can edit scripts.

**Review by parkervcp — CHANGES_REQUESTED** [review]

File: steamcmd_servers/space_engineers/egg-space-engineers.json (hunk 1 above)
Comment (parkervcp) [reply steamcmd_servers/space_engineers/egg-space-engineers.json:16]: The following can all be variables in the panel

WINEARCH=win64
WINEPREFIX=/home/container/.wine
WINEDEBUG=-all wine /home/container/spaceengineers/DedicatedServer64/SpaceEngineersDedicated.exe -path

Comment (TuEye) [reply steamcmd_servers/space_engineers/egg-space-engineers.json:16]: True. Didn't think about the fact that the use of an editable start script may have negative consequences.
With the variables it makes it even a bit nicer. Especially in relation to WINEDEBUG.

Just testing it through again and then pushing the changes in.

File: steamcmd_servers/space_engineers/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+### Disk
+This server uses about 7GB of diskspace.
+
+## Server Ports
+
+| Port  | default |
+|-------|---------|
+| Game  | 27016   |
+| Steam | 8766    |
+| API   | 8080    |
```
Comment (gOOvER) [inline steamcmd_servers/space_engineers/README.md:27]: The default Deamon Port is 8080 and if someone assign it, will break the deamon

**Comment by gOOvER** [comment]
What i also found; YOu use Wine5 in image and in installscript u install Wine4

`warning: Your version of wine 4.0 is no longer supported upstream. You should upgrade to 5.x`

**Comment by gOOvER** [comment]
Ok, for me the server dont Start. Stucks on:

```
Running on Debian 10.7
Current timezone: Etc/UTC
wine-6.0
steam user is not set.

Using anonymous user.

No appid set. Starting Server
First launch will throw some errors. Ignore them
:/home/container$ cd /home/container/spaceengineers/DedicatedServer64/; wine /home/container/spaceengineers/DedicatedServer64/SpaceEngineersDedicated.exe -path Z:\\home\\container\\config -noconsole -ignorelastsession
_XSERVTransmkdir: ERROR: euid != 0,directory /tmp/.X11-unix will not be created.
```

**Comment by gOOvER** [comment]
Server crash on Start:

```
Digest: sha256:133d470628ecff842859dbaf63f8a7137f89dc8251102448b09f9371395d577a 
Status: Image is up to date for quay.io/parkervcp/pterodactyl-images:debian_wine-5 
[Pterodactyl Daemon]: Finished pulling Docker container image
Running on Debian 10.7
Current timezone: Etc/UTC
wine-6.0
steam user is not set.

Using anonymous user.

No appid set. Starting Server
First launch will throw some errors. Ignore them
:/home/container$ cd /home/container/spaceengineers/DedicatedServer64/; wine /home/container/spaceengineers/DedicatedServer64/SpaceEngineersDedicated.exe -path Z:\\home\\container\\config -noconsole -ignorelastsession
_XSERVTransmkdir: ERROR: euid != 0,directory /tmp/.X11-unix will not be created.
2021-03-12 16:28:46.187: App Version: 01_197_180
2021-03-12 16:28:46.195: Is official: True [NO][NIS][NAMP]
2021-03-12 16:28:46.196: Environment.ProcessorCount: 8
2021-03-12 16:28:46.198: Environment.OSVersion: Microsoft Windows 7 Professional
 (Microsoft Windows NT 6.1.7601 Service Pack 1)
2021-03-12 16:28:46.198: Environment.CommandLine: "Z:\home\container\spaceengine
ers\DedicatedServer64\SpaceEngineersDedicated.exe" -path Z:\home\container\confi
g -noconsole -ignorelastsession
2021-03-12 16:28:46.199: Environment.Is64BitProcess: True
2021-03-12 16:28:46.199: Environment.Is64BitOperatingSystem: True
2021-03-12 16:28:46.199: Environment.Version: .NET Framework 4.8.3761.0
2021-03-12 16:28:46.200: Environment.CurrentDirectory: Z:\home\container\spaceen
gineers\DedicatedServer64
2021-03-12 16:28:46.200: CPU Info: Common KVM processor
2021-03-12 16:28:47.590: Bind IP : 0.0.0.0:27016
Setting breakpad minidump AppID = 244850
dbus[39]: arguments to dbus_connection_send_with_reply_and_block() were incorrect, assertion "connection != NULL" failed in file ../../../dbus/dbus-connection.c line 3544.
This is normally a bug in some application using the D-Bus library.

  D-Bus not built with -rdynamic so unable to print a backtrace
wine: Assertion failed at address 00007F1B87CC87BB (thread 0058), starting debugger...
wine client error:58: write: Bad file descriptor
2021-03-12 16:28:50.262: Networking service: Steam
2021-03-12 16:28:50.263: Server successfully started
2021-03-12 16:28:50.263: Product name: Space Engineers
2021-03-12 16:28:50.263: Desc: Space Engineers
2021-03-12 16:28:50.263: Public IP: 213.202.208.160
2021-03-12 16:28:50.264: User ID: [009...841]
2021-03-12 16:28:50.266: Server connected to Steam
2021-03-12 16:28:50.615: Server Name:
2021-03-12 16:28:50.615: World Name:
2021-03-12 16:28:50.632: Auto Restart Enabled
container@pterodactyl~ Server marked as offline...


[Pterodactyl Daemon]: ---------- Detected server process in a crashed state! ----------
[Pterodactyl Daemon]: Exit code: 255
[Pterodactyl Daemon]: Out of memory: false
[Pterodactyl Daemon]: Aborting automatic restart, last crash occurred less than 60 seconds ago.
```

File: steamcmd_servers/space_engineers/README.md (hunk 2 above)
Comment (TuEye) [reply steamcmd_servers/space_engineers/README.md:27]: Oops. Will change it. Thanks.

**Comment by TuEye** [comment]
> What i also found; YOu use Wine5 in image and in installscript u install Wine4
> 
> `warning: Your version of wine 4.0 is no longer supported upstream. You should upgrade to 5.x`
Yes, that's right. But unfortunately I had to do that. When I used a higher Wine version I had various problems when installing the Winetricks. Actually primarily in connection with .NET48. I tried with the current winehaq-stable and had no stable result.

Normally it should not be a problem. When starting the server, the wineprefix is automatically updated. That's why the first start takes a little longer.

**Review by parkervcp — CHANGES_REQUESTED** [review]
Why is there a zip file here?

**Comment by gOOvER** [comment]
> Why is there a zip file here?

It's the premade World; without it, Server cant start.

**Comment by gOOvER** [comment]
Please make this working with paths to your Reproi. SAo noone can test it really. I will not sowlload all and everythin manual.

When it's approved and ready to merge, you can change back the Links.

**Comment by TuEye** [comment]
Ok, just adjusted it

**Comment by gOOvER** [comment]
> Why is there a zip file here?

Thats the content of the zip File

chrome_7CbuycVfCx

**Review by gOOvER — APPROVED** [review]
After the Egg download all, i was able to connect to the Server. For me it works

**Comment by TuEye** [comment]
I would then reset the URL back to the original value. Or is there a reason that speaks against it?

**Comment by halkeye** [comment]
Is downloading a zip from a host the normal way to do it? Especially with the blanket master url its a bit scary, if someone changes something it could break installs for you, even if your setup didn't change.

Would it make more sense to md5 the files and embed them directly into the json?
CFG_MD5=$(cat SpaceEngineers-Dedicated.cfg| base64)

then in your script
```
if [ ! -f $HOME/config/SpaceEngineers-Dedicated.cfg ]; then echo $CFG_MD5 | base64 -d > $HOME/config/SpaceEngineers-Dedicated.cfg; fi
```

for example?

**Comment by TuEye** [comment]
To be honest, I can understand the concern. Having the file coded in the install script has a charm but also makes the script or this part rather unreadable.

Regarding the configuration file, I could imagine to simply place it in plain text in the Installscript and write it to a file. 

The start world on the other hand is rather difficult to provide in this way, it is simply too big. 

I could also remove the provision of the start world altogether.
Then I could note in the readme that it has to be uploaded by the user before the server can be started. Would that also be an option?

**Comment by parkervcp** [comment]
Putting them in the repo and pulling them on install is how I generally manage config files. This makes it easy to see publically that changes were made.

converting to MD5 and base64 obscures that a lot.

**Comment by halkeye** [comment]
👍 i was just throwing out ideas. As a random user i'd like the config as a heredoc instead of wget, and the save files optional, but as is its fine.

**Comment by TuEye** [comment]
For now, I have set the download link back to the pontetial new repo. 
If I should change something else, I can still do that.

**Pull request merged**